### PR TITLE
Backmerge 2024-07-16 (`nmdc-schema/main` -> `berkeley-schema-fy24/main`)

### PR DESCRIPTION
**Pull request by turbomam**
This is a merge of content present in `nmdc-schema` `main` into `berkeley-schema` `main` as of yesterday. It does not intentionally introduce any new content that wasn't already approved in the `nmdc-schema` PR approval process.

The NMDC PR template is included below but has not been filled out.

----

# Guidelines

## _Soft_ Schema Freeze

The `nmdc-schema` and `berkeley-schema-fy24` schemas are under a soft freeze, which means changes **should not** be made that have any downstream implications. To ensure this, all PRs created creating during the freeze will be closely reviewed with **every** component of the NMDC system in mind.

## Reviewers

To ensure no changes are made unexpectedly, PR creators will request reviews from _all_ [Berkeley Schema Roll Out task coordinators](https://docs.google.com/document/d/1XXN1YuaBuSkxPXeiLKm5YxYzXTamBPQrzzeLhlh7PWs/edit#heading=h.u52g8v319adh).

We expect task coordinators to review PRs and provide feedback/approval within 1 week of when they are identified as reviewers. 

PRs will **NOT** be merged until all task coordinators (or one of their delegates) have approved. 

Expedition, questions, and discussion can happen at any meeting.

Delays in review & merging should be addressed in meetings or with NMDC leadership.

# PR Information

## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation
- [ ] Schema change: Structure and content
  - created, updated, or deleted a `class`, `slot`, or `enum`
  - changed whether a `slot` is `multivalued`
  - changed the way a `slot` is assigned to a `class`
  - changed the `permissible_values` of an `enum`
  - _etc._
- [ ] Schema change: Cleanup and preparation
  - updated the description of a `class`, `slot`, or `enum`
  - updated the `mappings` of a `class`, `slot`, or `enum` to an ontology
  - added an `enum` for _future_ use (it is not in the `range` of any `slot`)
  - _etc._
     
## Description

> _PRs should be [small and concise](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/getting-started/best-practices-for-pull-requests#write-small-prs)._
>
> Aim to create small, focused pull requests that fulfill a single purpose. Smaller pull requests are easier and faster to review and merge, leave less room to introduce bugs, and provide a clearer history of changes.

- Replace this text with a description of what this PR branch contains. Please keep in mind that all reviewers will be reading this description. Example: "In this branch, I..."

## Related Issues

> All PRs should relate to or fix an issue(s). Please identify the issue(s) below.

- Related Issue(s): #
- Fixes: #

## Did you add/update any tests?

- [ ] Yes
- [ ] No _(Add a justification below)_
- [ ] I need help with writing tests

## Could this schema change make it so any valid data becomes invalid?

> This is a question about what the schema allows. It is not a question about what happens to exists in the NMDC database right now.
> 
> Example: If, in this PR branch, you renamed a `slot` from `foo` to `foo_bar`, the answer to this question would be "yes," **even if** nothing in the NMDC database _currently_ uses the `foo` `slot`.
>
> More examples: `slot` or `class` name changes, changes to a `slot`'s `multivalued` state, changes to a `slot`'s `range` (e.g. string to integer), changes to `slot` assignments to `class`es, changes to an `enum`'s `permissible_values`

- [ ] Yes _(A migrator is required)_
- [ ] No
- [ ] I need help determining this

## If you answered "Yes", does this PR branch include that migrator?

- [ ] Yes
- [ ] No, this PR is incomplete and I need help writing the migrator

## Does this PR have any downstream implications?

> Examples: any change here that requires a change to workflows, workflow automation, the Mongo-to-Postgres ingest process, Jupyter notebooks, the Runtime, etc.

- [ ] Yes _(Explain below)_
- [ ] No

### CODE REVIEW

**Comment by turbomam** [comment]
@aclum @mslarae13 @eecavanna here is a proposed backmerge for today's main branches

do you want me to fill out the PR template?

**Comment by turbomam** [comment]
Are we really going to wait for all task leaders to approve this? If so, we can't merge anything else into either branch until this is merge. Otherwise we'll have to start all over again.

**Comment by eecavanna** [comment]
During today's metadata meeting, I brought up the possibility of us (for future backmerge attempts) creating a branch off of `nmdc-schema/main` when starting the backmerge work, so that the backmerge is not impacted by subsequent changes on `nmdc-schema/main`.

I can elaborate here, on Slack, on Zoom, etc. or pair program with @turbomam during the next backmerge.

**Review by kheal — APPROVED** [review]

File: assets/yq-for-nmdc_schema_accepting_legacy_ids.txt
Comment (aclum) [inline assets/yq-for-nmdc_schema_accepting_legacy_ids.txt:1]: okay with this file since it matches nmdc-schema but I hope we don't need this file leagacy id related file anymore.

File: src/data/invalid/Database-WorkflowExecutionActivity-invalid.yaml
Comment (aclum) [inline src/data/invalid/Database-WorkflowExecutionActivity-invalid.yaml:5]:
```suggestion
    type: nmdc:WorkflowExecution
```

**Review by aclum — CHANGES_REQUESTED** [review]
See suggested change for invalid example data.

File: Makefile
Comment (eecavanna) [inline Makefile:92]: I added some indentation and put each filename on its own line (rather than 1 on one line, 2 on the next line [even though they are similar], 1 on the next line).

```suggestion
site: clean site-clean gen-project gendoc \
      nmdc_schema/gold-to-mixs.sssom.tsv \
      nmdc_schema/nmdc_materialized_patterns.schema.json \
      nmdc_schema/nmdc_materialized_patterns.yaml \
      migration-doctests
```

File: project.Makefile
Comment (eecavanna) [inline project.Makefile:111]: I'd prefer this be removed instead of commented out, but I don't consider it to be a deal-breaker on a "backmerge," given all the moving parts. It can be removed in a subsequent PR.

File: src/schema/nmdc.yaml
Comment (eecavanna) [inline src/schema/nmdc.yaml:986]: I'd prefer this be deleted instead of commented out, but I don't consider it to be a deal-breaker on a "backmerge," given the higher than usual amount of complexity involved in the merging task. It can be deleted in a subsequent commit/PR.

File: tests/test_json_schema_artifact.py
Comment (eecavanna) [inline tests/test_json_schema_artifact.py:29]: Props to whoever wrote these approachable docstrings! 👏

**Comment by eecavanna** [comment]
I left some suggestions in comments. I'll approve, given that none of them area deal breakers to me. Thanks for working on this, @turbomam!

**Review by eecavanna — APPROVED** [review]

File: Makefile
Comment (eecavanna) [reply Makefile:92]: It just occurred to me that lines like this often precede tab-indented lists of commands; in which case, this new indentation (I used spaces) might make the file more difficult to read instead of less difficult.

**Comment by mslarae13** [comment]
no concerns from me and no technical contribution

**Review by naglepuff — APPROVED** [review]
I don't think this will have an outsized impact (if any) on work being done for `nmdc-server`'s switch to be Berkeley compliant.

File: assets/yq-for-nmdc_schema_accepting_legacy_ids.txt
Comment (turbomam) [reply assets/yq-for-nmdc_schema_accepting_legacy_ids.txt:1]: really shouldn't be in either

File: src/data/invalid/Database-WorkflowExecutionActivity-invalid.yaml
Comment (turbomam) [reply src/data/invalid/Database-WorkflowExecutionActivity-invalid.yaml:5]: wow I'm surprised the example data files didn't catch that!

Comment (turbomam) [reply src/data/invalid/Database-WorkflowExecutionActivity-invalid.yaml:5]: maybe because we don't instantiate it? it's an abstract class?

File: Makefile
Comment (turbomam) [reply Makefile:92]: I agree that this is easier to read but I am hesitant to add white-space in a Makefile (which is already picky about XXX vs tabs). If we did that, I would want to apply it to all cases where a list of target prerequisites spills over the line on which the target is defined.

File: src/schema/nmdc.yaml
Comment (turbomam) [reply src/schema/nmdc.yaml:986]: believe it or not, there is a function, non-commented out copy of this in `src/schema/core.yaml`!

deleting this commend-out one as suggested

File: tests/test_json_schema_artifact.py
Comment (turbomam) [reply tests/test_json_schema_artifact.py:29]: @eecavanna at least in this branch, I am credited as the author of this entire file. I really don't think that's the case. Can you please help me figure out what happened?

**Review by aclum — APPROVED** [review]

File: Makefile
Comment (eecavanna) [reply Makefile:92]: Thanks for addressing this topic in conversation here. I'll resolve this conversation. I'm OK with proceeding with this snippet as it exists today.

File: tests/test_json_schema_artifact.py
Comment (eecavanna) [reply tests/test_json_schema_artifact.py:29]: According to Git, the file was created by you in this commit: https://github.com/microbiomedata/berkeley-schema-fy24/commit/e35178672aa559d4bfb8d24ee188e79bfde6d48c#diff-d66bad7d4246ba73f4053e7a121053a4c7898d20c631961595cc616d4077b75a

That is in contrast to Git saying that it existed before that commit.

I'll dig into this a little bit to come up with a hypothesis.

Comment (eecavanna) [reply tests/test_json_schema_artifact.py:29]: My guess is that _part_ of this merge was done by copy/pasting files or their content from one branch to the other, instead of using Git's own merge functionality. I don't know how likely that is, given the specific merge workflow you are using. 🤷

Comment (turbomam) [reply tests/test_json_schema_artifact.py:29]: thanks. if there's a way to restore credit/blame to @pkalita-lbl , I'd like to do that. I'm pretty sure I didn't edit the file.

**Review by mbthornton-lbl — APPROVED** [review]

**Pull request merged**
